util/dvsim/CircularList.py

```
@@ -0,0 +1,123 @@
+        self.count = 0
+
+    def __str__(self):
+        items = []
+        current = self.head.next_node
+        while current != self.head:
+            items.append(current.item.__str__())
+            current = current.next_node
+        return items.__str__()
+
+    def __iter__(self):
+        '''Iterate over the list, stopping at the last entry.'''
+        return CircularListIterator(self)
+
+    def __len__(self):
+        '''Returns the number of items in the list.'''
+        return self.count
+
+    def __bool__(self):
+        '''Returns True if there are items in the list, else False.'''
+        return bool(self.count)
+
+    def add(self, item):
+        '''Adds a new item at the end of the list.'''
+        node = Node(item=item, next_node=self.head)
+        self.tail.next_node = node
+        self.tail = node
+        self.count += 1
+
+    append = add
+    '''Alias of add().'''
+
+    def extend(self, items):
+        '''Extends the list with multiple items.'''
+        for item in items:
+            self.append(item)
+
+    def remove(self, item):
+        '''Removes the item from the list.'''
+        previous = self.head
+        current = self.head.next_node
+        while current != self.head:
+            if current.item == item:
+                previous.next_node = current.next_node
+                if self.tail == current:
+                    self.tail = previous
+                if self.ptr == current:
+                    self.ptr = previous
+                del current
+                self.count -= 1
+                return
+            previous = current
+            current = current.next_node
+        raise ValueError("List does not contain item {}".format(item))
+
+    def get_next_item(self):
+        '''Circularly iterate over the list and return an item.
+
+        This iteration loops back forever, returning an item if available.
+        Returns None if the list is empty.
```

Maybe raise an exception like `StopIteration` if the list is empty? Then we can store an `Optional[X]` in the circular list without any gymnastics.